### PR TITLE
Partition schema evolution for Parquet

### LINKED ISSUE #12212: Parquet Hive schema evolution checks based on column names
We occasionally run into an issue in our Hive parquet partitioned tables when users introduce new columns in the middle of the table. 
For example:
```
CREATE TABLE my_table ( 
    day varchar,                                           
    hour integer,                                          
    user_id_fast bigint)
WITH (                                                    
    format = 'PARQUET',                                    
    partitioned_by = ARRAY['day_part'])
;
```

If new columns are added in the middle:
```
CREATE TABLE my_table ( 
    day varchar,            
    is_attributed boolean,
    platform varchar,                               
    hour integer,                                          
    user_id_fast bigint)
...
;
```

Attempting to access one of the partitions with the older schema results in errors as Presto is matching the types of the columns in the partition and the table by index. So index 1 ends up being an integer in one and a boolean in the other which results in an error (https://github.com/prestodb/presto/blob/master/presto-hive/src/main/java/com/facebook/presto/hive/HiveSplitManager.java#L299). The behavior is similar in case of structs as well - https://github.com/prestodb/presto/blob/master/presto-hive/src/main/java/com/facebook/presto/hive/HiveCoercionPolicy.java#L99

From what I understand Hive does this checking based on columnNames in case of Parquet. Is this something we can implement here as well? (I'm not sure what the correct behavior should be for formats like ORC, so it might be that we need to have different checks for different formats). 

On similar lines - https://github.com/prestodb/presto/issues/8911. Once the flag `hive.parquet.use-column-names` is set, we use column names while reading Parquet data. This flag is currently not involved at the schema checking step.

### PR DESCRIPTION
**Pull request by imjalpreet**
Fixes #12212 

```
== RELEASE NOTES ==

Hive Changes
* Add support for partition schema evolution for parquet
```

### CODE REVIEW

**Comment by dborkar** [comment]
@aweisberg @zhenxiao can one of you please review this PR? thx

**Comment by zhenxiao** [comment]
thank you, @imjalpreet 
correct me if I am wrong. I think the problem could be resolved by enabling `use-column-names` by default

shall we backport:
https://github.com/trinodb/trino/pull/6479

**Comment by imjalpreet** [comment]
> thank you, @imjalpreet 
> 
> correct me if I am wrong. I think the problem could be resolved by enabling `use-column-names` by default
> 
> 
> 
> shall we backport:
> 
> https://github.com/trinodb/trino/pull/6479

@zhenxiao The changes I have made are required for supporting schema evolution at Partition level. Currently, we did not have that. Even if we enabled use-column-names it used to compare table and partition schema by index rather than names. 

Regarding the backport, yes we can next look at enabling it by default to match Hive's working. The trino PR you mentioned also contains changes for other formats ORC, so we should also work on adding support use-column-names in the case of ORC which is currently missing.

Let me know if you have any concerns.

**Comment by imjalpreet** [comment]
@zhenxiao an example error message can be seem below:

```
com.facebook.presto.spi.PrestoException: There is a mismatch between the table and partition schemas. The types are incompatible and cannot be coerced. The column 'a' in table 't' is declared as type 'string', but partition 'b=23/year=2020/dayofyear=16' declared column 'c' as type 'double'.
	at com.facebook.presto.hive.HiveSplitManager.lambda$getPartitionMetadata$4(HiveSplitManager.java:429)
	at com.google.common.collect.Iterators$6.transform(Iterators.java:785)
error
```

> was thinking use-column-names could handle table schema and parquet file schema evolution.

Yes it does handle that. But what it doesn't handle currently is table schema and partition schema evolution.

**Comment by zhenxiao** [comment]
get it.
are you porting:
https://github.com/trinodb/trino/commit/d30260f88ddfc381624daa3c5c4d0efdf306d744
https://github.com/trinodb/trino/commit/5790ed02c008d0e5ce18b21bb71f216fb44b8f3d
https://github.com/trinodb/trino/commit/385d826b15c005d3b16a9070eb342860aee86161
https://github.com/trinodb/trino/commit/a93b3b849a4af803e932000d4bfff0dc9691a0c9

could you please add Trino commits link in the commit message?

**Comment by zhenxiao** [comment]
get it. thank you, @imjalpreet I will take a look as soon
could you please follow the cherry-pick commit format? e.g. take a look at the commit message of: 83fe8e7c8aafabaa64d943038202bc02d26b326b

File: presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/MetastoreUtil.java [hunk 1]
```diff
@@ -313,21 +314,38 @@ public static Properties getHiveSchema(Map<String, String> serdeParameters, Map<
...
-            if (column == null) {
-                checkArgument(
-                        i < tableSchema.size(),
-                        "column descriptor for column with hiveColumnIndex %s not found: tableSchema: %s, partitionSchemaDifference: %s",
-                        i,
-                        tableSchema,
-                        partitionSchemaDifference);
-                column = tableSchema.get(i);
+
+        if (!tableToPartitionColumns.isPresent()) {
```
Comment (zhenxiao) [inline presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/MetastoreUtil.java:321]: how about simplify the logic:
if (tableToPartitionColumns.isPresent()) {
}
else {

File: presto-hive/src/main/java/com/facebook/presto/hive/HivePageSourceProvider.java [hunk 2]
```diff
@@ -652,14 +656,14 @@ public int getIndex()
         /**
          * @param columns columns that need to be returned to engine
          * @param requiredInterimColumns columns that are needed for processing, but shouldn't be returned to engine (may overlaps with columns)
-         * @param partitionSchemaDifference map from hive column index to hive type
+         * @param tableToPartitionMapping   table to partition mapping
```
Comment (zhenxiao) [inline presto-hive/src/main/java/com/facebook/presto/hive/HivePageSourceProvider.java:659]: reduce blank between tableToPartitionMapping and table

**Review by zhenxiao — APPROVED** [review]
looks good. 2 minor issues

File: presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/MetastoreUtil.java (hunk 1 above)
Comment (imjalpreet) [reply presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/MetastoreUtil.java]: Modified as requested

File: presto-hive/src/main/java/com/facebook/presto/hive/HivePageSourceProvider.java (hunk 2 above)
Comment (imjalpreet) [reply presto-hive/src/main/java/com/facebook/presto/hive/HivePageSourceProvider.java]: Done 👍🏼

**Comment by jiangjiguang** [comment]
> @zhenxiao an example error message can be seem below:
> 
> ```
> com.facebook.presto.spi.PrestoException: There is a mismatch between the table and partition schemas. The types are incompatible and cannot be coerced. The column 'a' in table 't' is declared as type 'string', but partition 'b=23/year=2020/dayofyear=16' declared column 'c' as type 'double'.
> 	at com.facebook.presto.hive.HiveSplitManager.lambda$getPartitionMetadata$4(HiveSplitManager.java:429)
> 	at com.google.common.collect.Iterators$6.transform(Iterators.java:785)
> error
> ```
> 
> > was thinking use-column-names could handle table schema and parquet file schema evolution.
> 
> Yes it does handle that. But what it doesn't handle currently is table schema and partition schema evolution.

@imjalpreet I don't understand a and c are the different columns, why have the error ?

**Comment by imjalpreet** [comment]
@jiangjiguang This can happen if order of the columns is different. Can you let me know if you have set `parquet.use-column-names` to true?

**Comment by jiangjiguang** [comment]
@imjalpreet What means "Partition schema evolution" ?   Can you give me a demo ?

**Pull request merged**